documentation/examples/custom-sd/adapter-usage/main.go

```diff
@@ -0,0 +1,155 @@
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package main
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"io/ioutil"
+	"net/http"
+	"os"
+	"time"
+
+	"github.com/go-kit/kit/log"
+	"github.com/prometheus/common/model"
+	"github.com/prometheus/prometheus/discovery/targetgroup"
+	"github.com/prometheus/prometheus/documentation/examples/custom-sd/adapter"
+	"gopkg.in/alecthomas/kingpin.v2"
+)
+
+var (
+	a             = kingpin.New("sd adapter usage", "Tool to generate file_sd config files for unimplemented SD mechanisms.")
+	outputFile    = a.Flag("output.file", "Output file for file_sd compatible file. Prints to stdout if none provided.").Default("/tmp/custom_sd.json").String()
+	listenAddress = a.Flag("listen.address", "The address the http sd is listening on for requests.").String()
+	logFile       = a.Flag("log.file", "Location of logfile.").String()
+	logger        log.Logger
+)
+
+type sdConfig struct {
+	listenAddress   string
+	refreshInterval int
+	name            string
+}
+
+type discovery struct {
+	listenAddress   string
+	refreshInterval time.Duration
+	name            string
+}
+
+type targetsResponse struct {
+	services map[string][]string
+}
+
+func (d *discovery) jsonToTargets(data []byte) ([]*targetgroup.Group, error) {
+	var arbJSON interface{}
+	var tgs []*targetgroup.Group
+
+	err := json.Unmarshal(data, &arbJSON)
+	if err != nil {
+		return nil, err
+	}
+
+	targetsMap := arbJSON.(map[string]interface{})
+	for k, v := range targetsMap {
+		targets := v.([]interface{})
+		targetsLabels := []model.LabelSet{}
+		tg := targetgroup.Group{Source: d.name}
+		for _, target := range targets {
+			targetLabel := model.LabelSet{
+				model.AddressLabel: model.LabelValue(target.(string)),
+			}
+			targetsLabels = append(targetsLabels[:], targetLabel)
+		}
+		labels := model.LabelSet{}
+		labels["dc"] = model.LabelValue(k)
+		tg.Labels = labels
+		tg.Targets = targetsLabels
+		// for our http sd we use the dc name as the source
+		tg.Source = k
+		tgs = append(tgs, &tg)
+	}
+	return tgs, nil
+}
+
+func (d *discovery) Run(ctx context.Context, ch chan<- []*targetgroup.Group) {
+	targetsURL := fmt.Sprintf("http://%s/targets", d.listenAddress)
+
+	// send the target groups down the channel
+	for c := time.Tick(d.refreshInterval); ; <-c {
+		resp, err := http.Get(targetsURL)
+		if err != nil {
+			// log
+			continue
+		}
+		body, err := ioutil.ReadAll(resp.Body)
+		if err != nil {
+			// log
+			continue
+		}
+
+		tgs, err := d.jsonToTargets(body)
+		if err != nil {
+			logger.Log("error", err.Error())
+		} else {
+			ch <- tgs
+		}
+	}
+}
+
+func newDiscovery(cfg sdConfig) *discovery {
+	return &discovery{
+		listenAddress:   cfg.listenAddress,
+		refreshInterval: time.Second * time.Duration(cfg.refreshInterval),
+	}
+}
+
+func main() {
```

Comment:
Why do you have two main functions?

Reply:
I wrote the example as two executables, `adapter-usage/main.go` is the executable that queries the SD implementation and turns it's target groups into JSON to write for file_sd, `http-discovery/main.go` is the executable that runs this mock HTTP SD.